### PR TITLE
phase 1, migrate the outer shell of cosign to cobra

### PR DESCRIPTION
**Pull request by n3wscott**
#### Summary

This is the first in several PRs to migrate cosign to cobra. In this PR I have focused on:

- Adding deprecation warnings for using non-POSIX compliant flags. 
- grouping subcommands together where needed to prevent circular dependencies.
- moving core non-cli logic out of `cmd/` and into `pkg/` where needed to prevent circular dependencies.

#### Ticket Link

Related to #706 

#### Release Note

```release-note
- Deprecation Warning: cosign is moving to POSIX style flags. All flags that are single dash and a word are deprecated in favor of double dash and word (`-foo` to `--foo`) and double dash single letter flags are deprecated in favor of single dash flags (`--f` to `-f`).
```

### CODE REVIEW

File: cmd/cosign/cli/options/root.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+type RootOptions struct {
+	OutputFile string
+	Verbose    bool
+}
+
+func AddRootArgs(cmd *cobra.Command, o *RootOptions) {
+	cmd.PersistentFlags().StringVar(&o.OutputFile, "output-file", "",
+		"log output to a file")
+
+	cmd.PersistentFlags().BoolVarP(&o.Verbose, "verbose", "d", false,
```
Comment (dekkagaijin) [inline cmd/cosign/cli/options/root.go:32]:
```suggestion
	cmd.PersistentFlags().BoolVarP(&o.Verbose, "verbose", "v", false,
```

File: cmd/cosign/cli/options/root.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+// RootOptions define flags and options for the root cosign cli.
+type RootOptions struct {
+	OutputFile string
+	Verbose    bool
+}
+
+func AddRootArgs(cmd *cobra.Command, o *RootOptions) {
+	cmd.PersistentFlags().StringVar(&o.OutputFile, "output-file", "",
```
Comment (dekkagaijin) [inline cmd/cosign/cli/options/root.go:29]: `--log-out`? `--log-file`?

We'll probably be wanting to use `--output-file` and/or `-o` for the commands that rewrite dockerfiles and k8s manifests

File: cmd/cosign/cli/sign/sign_blob.go [hunk 3]
```diff
@@ -77,9 +80,9 @@ EXAMPLES
...
 		Exec: func(ctx context.Context, args []string) error {
 			// A key file is required unless we're in experimental mode!
-			if !EnableExperimental() {
-				if !oneOf(*key, *sk) {
-					return &KeyParseError{}
+			if !options.EnableExperimental() {
+				if !options.OneOf(*key, *sk) {
+					return &options.KeyParseError{}
 				}
 			}
```
Comment (dekkagaijin) [inline cmd/cosign/cli/sign/sign_blob.go:87]:
```suggestion
			if !options.EnableExperimental() && !options.OneOf(*key, *sk) {
				return &options.KeyParseError{}
			}
```

File: cmd/cosign/cli/verify/verify_blob.go [hunk 4]
```diff
@@ -114,19 +118,19 @@ func isb64(data []byte) bool {
 	return err == nil
 }
 
-func VerifyBlobCmd(ctx context.Context, ko KeyOpts, certRef, sigRef, blobRef string) error {
-	var pubKey signature.Verifier
+func VerifyBlobCmd(ctx context.Context, ko sign.KeyOpts, certRef, sigRef, blobRef string) error {
```
Comment (dekkagaijin) [inline cmd/cosign/cli/verify/verify_blob.go:121]:
```suggestion
func Blob(ctx context.Context, ko sign.KeyOpts, certRef, sigRef, blobRef string) error {
```

File: cmd/cosign/cli/verify/verify_blob.go [hunk 5]
```diff
@@ -32,12 +32,16 @@ import (
...
+	sigs "github.com/sigstore/cosign/pkg/signature"
 	rekorClient "github.com/sigstore/rekor/pkg/client"
 	"github.com/sigstore/sigstore/pkg/cryptoutils"
-	"github.com/sigstore/sigstore/pkg/signature"
-	"github.com/sigstore/sigstore/pkg/signature/options"
+	sigstoresigs "github.com/sigstore/sigstore/pkg/signature"
+	signatureoptions "github.com/sigstore/sigstore/pkg/signature/options"
 )
 
 func VerifyBlob() *ffcli.Command {
```
Comment (dekkagaijin) [inline cmd/cosign/cli/verify/verify_blob.go:47]:
```suggestion
func BlobCmd() *ffcli.Command {
```

File: cmd/cosign/cli/verify/verify.go [hunk 6]
```diff
@@ -27,15 +27,19 @@ import (
...
 	"github.com/sigstore/cosign/pkg/cosign"
 	"github.com/sigstore/cosign/pkg/cosign/pivkey"
+	"github.com/sigstore/cosign/pkg/image"
+	sigs "github.com/sigstore/cosign/pkg/signature"
 	"github.com/sigstore/sigstore/pkg/signature"
 	"github.com/sigstore/sigstore/pkg/signature/payload"
 )
 
 // VerifyCommand verifies a signature on a supplied container image
 type VerifyCommand struct {
```
Comment (dekkagaijin) [inline cmd/cosign/cli/verify/verify.go:41]: This one's weirder than the other surfaces, but would be good to address the linter feedback here, too

File: cmd/cosign/cli/options/root.go (hunk 1 above)
Comment (n3wscott) [reply cmd/cosign/cli/options/root.go:32]: the current option is `d` for debug, but debug is not a normal flag, verbose is :idk:

File: cmd/cosign/cli/options/root.go (hunk 2 above)
Comment (n3wscott) [reply cmd/cosign/cli/options/root.go:29]: I don't want to change the flag, but yeah I like that better. Let's get the refactor in place and revisit what we want long term

File: cmd/cosign/cli/verify/verify_blob.go (hunk 4 above)
Comment (n3wscott) [reply cmd/cosign/cli/verify/verify_blob.go:123]: :D I started doing this but all of the files are not constant. the plan is to move all of this code into another form, so this whole thing is throw away anyway soon. let's hold off renaming at this moment, I added a `// nolint` comment for now.

File: cmd/cosign/cli/verify/verify.go (hunk 6 above)
Comment (n3wscott) [reply cmd/cosign/cli/verify/verify.go:42]: yeah, agreed, but not in this PR?

File: cmd/cosign/cli/verify/verify_blob.go (hunk 4 above)
Comment (dekkagaijin) [reply cmd/cosign/cli/verify/verify_blob.go:123]: ty

File: cmd/cosign/cli/verify/verify.go (hunk 6 above)
Comment (dekkagaijin) [reply cmd/cosign/cli/verify/verify.go:41]: sgtm

**Review by dekkagaijin — APPROVED** [review]
approved w/ comments for further iteration

**Pull request merged**